go/felix/intdataplane/ipip_mgr_test.go

```diff
@@ -0,0 +1,444 @@
+				Expect(dataplane.RunCmdCalled).To(BeFalse())
+			})
+			It("should avoid setting the interface UP again", func() {
+				Expect(dataplane.LinkSetUpCalled).To(BeFalse())
+			})
+			It("should avoid setting the MTU again", func() {
+				Expect(dataplane.LinkSetMTUCalled).To(BeFalse())
+			})
+			It("should avoid setting the address again", func() {
+				Expect(dataplane.AddrUpdated).To(BeFalse())
+			})
+		})
+
+		Describe("after second call with different params", func() {
+			BeforeEach(func() {
+				dataplane.ResetCalls()
+				ipipMgr.configureIPIPDevice(1500, ip2)
+			})
+			It("should avoid creating the interface", func() {
+				Expect(dataplane.RunCmdCalled).To(BeFalse())
+			})
+			It("should avoid setting the interface UP again", func() {
+				Expect(dataplane.LinkSetUpCalled).To(BeFalse())
+			})
+			It("should set the MTU", func() {
+				Expect(dataplane.tunnelLinkAttrs.MTU).To(Equal(1500))
+			})
+			It("should reconfigure the address", func() {
+				Expect(dataplane.addrs).To(HaveLen(1))
+				Expect(dataplane.addrs[0].IP.String()).To(Equal("10.0.0.2"))
+			})
+		})
+
+		Describe("after second call with nil IP", func() {
+			BeforeEach(func() {
+				dataplane.ResetCalls()
+				ipipMgr.configureIPIPDevice(1500, nil)
+			})
+			It("should avoid creating the interface", func() {
+				Expect(dataplane.RunCmdCalled).To(BeFalse())
+			})
+			It("should avoid setting the interface UP again", func() {
+				Expect(dataplane.LinkSetUpCalled).To(BeFalse())
+			})
+			It("should set the MTU", func() {
+				Expect(dataplane.tunnelLinkAttrs.MTU).To(Equal(1500))
+			})
+			It("should remove the address", func() {
+				Expect(dataplane.addrs).To(HaveLen(0))
+			})
+		})
+	})
+
+	Describe("after calling configureIPIPDevice with no IP", func() {
+		BeforeEach(func() {
+			ipipMgr.configureIPIPDevice(1400, nil)
+		})
+
+		It("should create the interface", func() {
+			Expect(dataplane.tunnelLink).ToNot(BeNil())
+		})
+		It("should set the MTU", func() {
+			Expect(dataplane.tunnelLinkAttrs.MTU).To(Equal(1400))
+		})
+		It("should set the interface UP", func() {
+			Expect(dataplane.tunnelLinkAttrs.Flags).To(Equal(net.FlagUp))
+		})
+		It("should configure the address", func() {
+			Expect(dataplane.addrs).To(HaveLen(0))
+		})
+	})
+
+	// Cover the error cases.  We pass the error back up the stack, check that that happens
+	// for all calls.
+	expNumCalls := 9
+	It("a successful call should only call into dataplane expected number of times", func() {
+		Expect(dataplane.NumCalls).To(BeNumerically("<=", expNumCalls))
+	})
+	for i := 1; i < expNumCalls; i++ {
+		if i == 1 {
+			continue // First LinkByName failure is handled.
+		}
+		i := i
+		Describe(fmt.Sprintf("with a failure after %v calls", i), func() {
+			BeforeEach(func() {
+				dataplane.ErrorAtCall = i
+			})
+
+			It("should return the error", func() {
+				Expect(ipipMgr.configureIPIPDevice(1400, ip)).To(Equal(mockFailure))
+			})
+
+			Describe("with an IP to remove", func() {
```

Comment:
Yes, but that code (after the failure case I'm trying to hit) is common so I don't think it matters too much